src/base/associative_cache.hh

```diff
@@ -0,0 +1,308 @@
+
+        initParams(numEntries, associativity);
+    }
+
+    /**
+     * Get the tag for the addr
+     * @param addr Addr to get the tag for
+     * @return Tag for the address
+     */
+    virtual Addr getTag(const Addr addr) const
+    {
+        return indexingPolicy->extractTag(addr);
+    }
+
+    /**
+     * Do an access to the entry if it exists.
+     * This is required to update the replacement information data.
+     * @param addr key to the entry
+     * @return The entry if it exists
+     */
+    virtual Entry* accessEntry(const Addr addr)
+    {
+        auto entry = findEntry(addr);
+
+        if (entry) {
+            replPolicy->touch(entry->replacementData);
+        }
+
+        return entry;
+    }
+
+    /**
+     * Update the replacement information for an entry
+     * @param Entry to access and upate
+     */
+    virtual void accessEntry(Entry *entry)
+    {
+        replPolicy->touch(entry->replacementData);
+    }
+
+    /**
+     * Find an entry within the set
+     * @param addr key element
+     * @return returns a pointer to the wanted entry or nullptr if it does not
+     *  exist.
+     */
+    virtual Entry* findEntry(const Addr addr) const
+    {
+        auto tag   = getTag(addr);
+
+        auto entries = indexingPolicy->getPossibleEntries(addr);
+
+        for (auto it = entries.begin(); it != entries.end(); it++) {
+            Entry *entry = static_cast<Entry*>(*it);
+            if (entry->matchTag(tag)) {
+                return entry;
+            }
+        }
+
+        return nullptr;
+    }
+
+    /**
+     * Find a victim to be replaced
+     * @param addr key to select the possible victim
+     * @result entry to be victimized
+     */
+    virtual Entry* findVictim(const Addr addr)
+    {
+        auto candidates = indexingPolicy->getPossibleEntries(addr);
+
+        auto victim = static_cast<Entry*>(replPolicy->getVictim(candidates));
+
+        invalidate(victim);
+
+        return victim;
+    }
+
+    /**
+     * Invalidate an entry and its respective replacement data.
+     *
+     * @param entry Entry to be invalidated.
+     */
+    virtual void invalidate(Entry *entry)
+    {
+        entry->invalidate();
+        replPolicy->invalidate(entry->replacementData);
+    }
+
+    /**
+     * Indicate that an entry has just been inserted
+     * @param addr key of the container
+     * @param entry pointer to the container entry to be inserted
+     */
+    virtual void insertEntry(Addr addr, Entry *entry)
+    {
+        entry->insert(indexingPolicy->extractTag(addr));
+        replPolicy->reset(entry->replacementData);
+    }
+
+    /**
+     * Find the set of entries that could be replaced given
+     * that we want to add a new entry with the provided key
+     * @param addr key to select the set of entries
+     * @result vector of candidates matching with the provided key
+     */
+    std::vector<Entry *>
+    getPossibleEntries(const Addr addr) const
```

Comment:
Why is this new function being added? I don´t think it is being used anywhere in this PR. If this function is eventually needed to make something else use the cache (e.g., what Giacomo requests), it should be added in the respective PR that makes use of it. This approach of not adding until is needed is the standard to avoid having stale/forgotten code.